src/main/jbake/content/blog/2016/12/19/hawkular-metrics-compression.adoc

```diff
@@ -0,0 +1,93 @@
+
+At this point we now have two distinct streams, one with `DataPoint<T>` items and one with `ByteBuffers`. The first job is obviously to transform the `Observable<ByteBuffer>` to `Observable<DataPoint<T>>` in the given sorting order. For this, we have a transformer that takes the input ByteBuffer from the Cassandra driver, reads the first byte as our header and checks then what to do. Given that we only support a single compression method in this release, the straightforward approach is to then decompress the byte array and transform the compressed block back to uncompressed DataPoints. As we already know the type of the metric that's going to be read from the compressed block, we can correctly read either `long` or a `double` and as such avoid any casting that would reduce precision.
+
+Next, we must read from these two streams in the given order as the datapoints could be interleaved. Luckily, for this to function we don't need to do in-memory sorting of everything, but instead we read one point from each of the streams and then with a comparator select which one to forward and which one will be hold in the memory for next comparison. And then we request the next one from the Cassandra / uncompressed pool and continue. This allows us to reduce the memory pressure by keeping only one active datapoint and one uncompressed block in the memory for read operations. If either stream ends, then the other stream is allowed to push data without comparing it.
+
+This way, all the internal calls to getting datapoints (which go through a single method, link:github[findDataPoints()]) are unaware if the datapoints are coming from compressed or uncompressed streams. This allows easier refactoring in the future where we could read from several different streams if we so decided and no changes to elsewhere in the application is required.
+
+anchor:benchmark[]
+
+=== Benchmarks
+
+Disclaimer: the following performance numbers are an indication of results with the given test data. Your real world results can vary.
+
+The results are created against Cassandra 3.7, using LZ4 table compression settings.
+
+The dataset consist of real world data coming from a Windows host. A total of 19166 different time series are stored in the dataset, with each serie having a 10910 measurements with a millisecond precision. That means the total is 209110368 (209M) datapoints. All of them were stored as gauge type with 64 bit floating point precision.
+
+[cols="2*", options="header"]
+.Storage usage, data set #1
+|===
+|0.19.0
+|0.22.0
+
+| 3.2GB
+| 158MB
+|===
+
```

Comment:
> This allows us to reduce the memory pressure by keeping only one active datapoint and one uncompressed block in the memory for read operations.

This is not really accurate. We keep at  minimum a query result set page in memory at a time. We use a default value of 1,000. And I say minimum because I do not know if we do anything to aggressively release pages/rows that we are done with.